mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -0,0 +1,727 @@
+        let timer = Instant::now();
+        let data = binary_serialize(publication)?;
+        let data = Data::new(&data);
+        let data_size = binary_serialize_size(&data)?;
+        let data_hash = calculate_hash(&data);
+
+        let write_index = self.pointers.write;
+        let read_index = self.pointers.read;
+        let order = self.order;
+        let key = write_index;
+
+        let block_header =
+            BlockHeaderWithHash::new(data_hash, data_size, order, read_index, write_index);
+        let block_size = serialized_block_size()?;
+
+        let total_size = block_size + data_size;
+        // Check to see if we might corrupt data if we write, if so, return err of full.
+        // There are two cases for this:
+        // 1. write has wrapped around and is now behind read
+        if write_index < read_index && write_index + total_size > read_index {
+            return Err(StorageError::RingBuffer(RingBufferError::Full));
+        }
+        // 2. write has reached the end (or close enough) but read has not moved
+        if read_index == 0 && write_index + total_size > self.max_file_size {
+            return Err(StorageError::RingBuffer(RingBufferError::Full));
+        }
+
+        let mut start = write_index;
+
+        // Check if an existing block header is present. If the block is there
+        // and if the overwrite flag is not set then we shouldn't write.
+        let result = load_block_header(self.mmap, start, block_size, self.max_file_size);
+        if let Ok(block_header) = result {
+            let should_not_overwrite = block_header.inner().should_not_overwrite();
+            if should_not_overwrite {
+                return Err(StorageError::RingBuffer(RingBufferError::Full));
+            }
+        }
+
+        let should_flush = self.should_flush();
+
+        save_block_header(
+            self.mmap,
+            &block_header,
+            start,
+            self.max_file_size,
+            should_flush,
+        )?;
+
+        let mut end = start + block_size;
+        start = end % self.max_file_size;
+
+        save_data(self.mmap, &data, start, self.max_file_size, should_flush)?;
+
+        end = start + data_size;
+        self.pointers.write = end % self.max_file_size;
+
+        self.wake_up_task();
+
+        self.order += 1;
+
+        self.flush_state
+            .update(1, total_size, timer.elapsed().as_millis() as usize);
+        if should_flush {
+            self.flush_state.reset(&self.flush_options);
+        }
+
+        Ok(Key { offset: key as u64 })
+    }
+
+    fn batch(&self, count: usize) -> StorageResult<VecDeque<(Key, Publication)>> {
+        let write_index = self.pointers.write;
+        let read_index = self.pointers.read;
+        let block_size = serialized_block_size()?;
+
+        // If read would go into where writes are happening then we don't have data to read.
+        if read_index == write_index {
```

Comment:
what if we filled up the whole buffer exactly, so write_index wrapped around and become equal to read_index again? 
- do we need to keep "wrapped" flag?
- or we should disallow write_index to wrap and reach read_index by checking `write_index < read_index && write_index + total_size **>=** read_index`

Reply:
* Good catch! Let me add a test for this very situation and it would appear some 'wrapped' flag might be necessary or as you said if we reach the max_file_size with the write, it should stay there until the next write. This combined with the change mentioned above in insert on correcting the bound checking with a write should work. Either case, more tests!
* I think write should still be allowed to wrap and reach read. That way we can at least optimize for space.